File: apis/city-tasks-api/src/main/java/com/hiperium/city/tasks/api/models/Task.java
package com.hiperium.city.tasks.api.models;

import com.hiperium.city.tasks.api.utils.enums.EnumDeviceOperation;
import com.hiperium.city.tasks.api.utils.enums.EnumTaskStatus;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.ZonedDateTime;

@Data
@Builder
@Entity
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "HIP_CTY_TASKS")
public class Task {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "HIP_CTY_TASKS_SEQ")
    @SequenceGenerator(name = "HIP_CTY_TASKS_SEQ", sequenceName = "HIP_CTY_TASKS_SEQ", allocationSize = 1)
    private Long id;

    @Column(name = "name", length = 60, nullable = false)
    private String name;

    @Column(name = "description")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 3, nullable = false)
    private EnumTaskStatus status;

    @Column(name = "job_id", length = 30, nullable = false)
    private String jobId;

    @Column(name = "device_id", length = 30, nullable = false)
    private String deviceId;

    @Enumerated(EnumType.STRING)
    @Column(name = "device_operation", length = 10, nullable = false)
    private EnumDeviceOperation deviceOperation;

    @Column(name = "device_execution_command", length = 90)
    private String deviceExecutionCommand;

    @Column(name = "task_hour", nullable = false)
    private Integer hour;

    @Column(name = "task_minute", nullable = false)
    private Integer minute;

    @Column(name = "execution_days", length = 30, nullable = false)
    private String executionDays;

    @Column(name = "execute_until")
    private ZonedDateTime executeUntil;

    @Column(name = "created_at", nullable = false)
    private ZonedDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private ZonedDateTime updatedAt;
}

File: apps/city-tasks-app/src/app/pages/tasks/model/task.ts
export interface Task {
  id: number;
  name: string;
  description?: string;
  deviceId: string;
  deviceOperation?: string;
  hour: number;
  minute: number;
  executionDays: string;
  executionCommand: string;
  executeUntil: number;
  createdAt: number;
  updatedAt: number;
}

export const compareTasks = (t1: Task, t2: Task): number => {
  const result = t1.hour - t2.hour;
  if (result === 0) {
    const result2 = t1.minute - t2.minute;
    return taskComparatorResult(result2);
  } else {
    return taskComparatorResult(result);
  }
};

const taskComparatorResult = (result: number): number => {
  if (result > 0) {
    return 1;
  } else if (result < 0) {
    return -1;
  } else {
    return 0;
  }
};
